test/extended/oauth/redirect_uris.go

```diff
@@ -0,0 +1,148 @@
+package oauth
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+
+	"github.com/RangelReale/osincli"
+	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	configv1 "github.com/openshift/api/config/v1"
+	oauthv1 "github.com/openshift/api/oauth/v1"
+	exutil "github.com/openshift/origin/test/extended/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+)
+
+type testCaseRedirectURIs struct {
+	client   *oauthv1.OAuthClient
+	endpoint string
+	uri      string
+
+	expectedCode      int
+	responseValidator func(*http.Request, *http.Response) bool
+}
+
+var _ = g.Describe("[sig-auth][Feature:OAuthServer]", func() {
+	g.Describe("OAuthClientWithRedirectURIs", func() {
+		ctx := context.Background()
+		oc := exutil.NewCLI("oauthclient-with-redirect-uris")
+		httpClient := &http.Client{
+			Transport: &http.Transport{
+				Proxy:           http.ProxyFromEnvironment,
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			},
+			// special case to not follow redirects; rather return most recent response
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
+			},
+		}
+
+		g.It("must validate request URIs according to oauth-client definition", func() {
+
+			g.By("check oauth-openshift routes")
+			controlPlaneTopology, err := exutil.GetControlPlaneTopology(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			if *controlPlaneTopology == configv1.ExternalTopologyMode {
+				e2eskipper.Skipf("External clusters do not contain oauth-openshift routes")
+			}
+
+			g.By("create oauth-clients")
+			loopbackClient, err := generateRedirectURIOAuthClient(ctx, oc, "oauth-client-with-loopback-uris", []string{"http://127.0.0.1/callback", "http://[::1]/callback"})
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(loopbackClient).NotTo(o.BeNil())
+			nonLoopbackClient, err := generateRedirectURIOAuthClient(ctx, oc, "oauth-client-with-non-loopback-uris", []string{"https://www.redhat.com"})
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(nonLoopbackClient).NotTo(o.BeNil())
+
+			oauthDiscoveryData := getOAuthWellKnownData(oc)
+			challenge, method, _, err := osincli.GeneratePKCE()
+
+			bodyValidator := func(_ *http.Request, response *http.Response) bool {
```

Comment:
`bodyValidator` implies the assigned function returns a validator but it returns a `bool`. suggestion: let's just call this `validateBody`.